File: JavaScript/13_RomanToInteger.js
/**
 * @param {string} s
 * @return {number}
 */
const romanToInt = function(s) {
    const letterValues = {
        I: 1,
        V: 5,
        X: 10,
        L: 50,
        C: 100,
        D: 500,
        M: 1000
    };
    const reverseS = s.split('').reverse();
    let total = 0;

    reverseS.forEach(letter => {
       total += letterValues[letter];
       if (letter === "I" && total >= 5) {
           total -= 2;     
       } else if (letter === "X" && total >= 50) {
           total -= 20;
       } else if (letter === "C" && total >= 500) {
           total -= 200;
       }
    })
    return total;
}; 

const testCases = [
    {
        s: "III",
        expected: 3
    },
    {
        s: "LVIII",
        expected: 58
    },
    {
        s: "MCMXCIV",
        expected: 1994
    },
]

testCases.forEach((testCase, i) => {
    const output = romanToInt(testCase.s);
    if (JSON.stringify(output) === JSON.stringify(testCase.expected)) {
        console.log(`Test Case ${i}: Passed. Expected: ${JSON.stringify(testCase.expected)}, Output: ${JSON.stringify(output)} `)
    } else {
        console.log(`Test Case ${i}: Failed. Expected: ${JSON.stringify(testCase.expected)}, Output: ${JSON.stringify(output)} `)
    }
})

File: CSharp/13_RomanToInteger.cs
namespace LeetCodeSolutions {
    public static class RomanToInteger {
       public static int Solution(string s) {
           Dictionary<char, int> letterValues = new Dictionary<char, int> {
               {'I', 1},
               {'V', 5},
               {'X', 10},
               {'L', 50},
               {'C', 100},
               {'D', 500},
               {'M', 1000}
           };

           char[] stringArray = s.ToCharArray();
           Array.Reverse(stringArray);
           string sReversed = new string(stringArray); 
           int total = 0;

           for (int i = 0; i < sReversed.Length; i++) {
               total += letterValues[sReversed[i]];
               if (sReversed[i] == 'I' && total >= 5) {
                   total -= 2;
               } else if (sReversed[i] == 'X' && total >= 50) {
                   total -= 20; 
               } else if (sReversed[i] == 'C' && total >= 500) {
                   total -= 200; 
               }
           }
           return total; 
       }

       public static void RunTests() {
            Console.WriteLine("Running Roman To Integer Tests...");

            var testCases = new List<(string, int)> {
                ("III", 3),
                ("LVIII", 58),
                ("MMMDCCXLIX", 3749),
                ("MCDLXIX", 1469),
                ("MCMXCIV", 1994)
            };

            foreach (var (s, expected) in testCases) {
                try {
                    var result = Solution(s);
                    if (result == expected) {
                        Console.WriteLine(
                            $"Test Passed: Input: {s}, Expected: {expected}, Result: {result}"
                        );
                    } else {
                        Console.WriteLine(
                            $"Test Falied: Input: {s}, Expected: {expected}, Result: {result}"
                        );
                    }
                } catch (Exception ex) {
                    Console.WriteLine($"Test Failed: {ex.Message}");
                }
            }
       } 
    }
}